### PR TITLE
feat(connector): [Nexixpay] add mandates flow for cards

### PR DESCRIPTION
**Pull request by mrudulvajpayee4935**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Integrate mandates flow for card payments.
https://developer.nexi.it/en/funzionalita-/pagamenti-ricorrenti

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

### Cypress test:
**1) Multiuse mandates**
Screenshot 2024-11-29 at 12 50 53 PM

**2) Singleuse mandates**
Screenshot 2024-11-29 at 12 51 05 PM

**3) Test cases in the headless mode**
Screenshot 2024-12-04 at 11 58 30 AM

**Note:**  
1. Skipping no3ds test cases as it is not supported by nexixpay
2. Remaining few test cases in headless mode is failing due to inconsistency in test environment from the connector - cc: @likhinbopanna @Gnanasundari24 **
### Create Call
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_VyjzlxA4qJnZUAalRRpB8MGDRzw6moojPrhbHThoGwxrZiyCvNnuIdVKoWvumalE' \
--data '{
    "amount": 3545,
    "currency": "EUR",
    "confirm": false,
    "customer_id": "tester799",
    "setup_future_usage": "off_session"
}'
```
### Confirm Call
```
curl --location 'http://localhost:8080/payments/pay_sFK1K3juhPteyDXgEH8X/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_VyjzlxA4qJnZUAalRRpB8MGDRzw6moojPrhbHThoGwxrZiyCvNnuIdVKoWvumalE' \
--data-raw '{
    "payment_method": "card",
    "return_url": "https://google.com",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4349940199004549",
            "card_exp_month": "12",
            "card_exp_year": "30",
            "card_holder_name": "Joseph Does",
            "card_cvc": "396"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "IT",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "IT",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "customer_acceptance": {
        "acceptance_type": "offline"
    }
}'
```
### MIT Call
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_VyjzlxA4qJnZUAalRRpB8MGDRzw6moojPrhbHThoGwxrZiyCvNnuIdVKoWvumalE' \
--data '{
    "amount": 3545,
    "currency": "EUR",
    "confirm": true,
    "customer_id": "tester799",
    "recurring_details": {
        "type": "payment_method_id",
        "data": "pm_WqDpelm1rZ4yvUkf579X"
    },
    "off_session": true
}'
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/hyperswitch_connectors/src/connectors/nexixpay.rs [hunk 1]
```diff
@@ -465,17 +480,31 @@ impl ConnectorIntegration<Authorize, PaymentsAuthorizeData, PaymentsResponseData
...
-            .parse_struct("NexixpayPaymentsResponse")
-            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
-        event_builder.map(|i| i.set_response_body(&response));
-        router_env::logger::info!(connector_response=?response);
-        RouterData::try_from(ResponseRouterData {
-            response,
-            data: data.clone(),
-            http_code: res.status_code,
-        })
+        if data.request.connector_mandate_id().is_none() {
```
Comment (awasthi21) [inline crates/hyperswitch_connectors/src/connectors/nexixpay.rs:483]: this check won't be needed if we use an enum for the PaymentsResponse
`enum NexixpayResponse {
    PaymentsPresonse(NexixpayPaymentsResponse),
    MandateResponse(NexixpayMandateResponse),
}`
`

Comment (awasthi21) [reply crates/hyperswitch_connectors/src/connectors/nexixpay.rs:483]: Take reference from Fiuu: Transformers
```#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum FiuuPaymentsResponse {
    PaymentResponse(Box<PaymentsResponse>),
    QRPaymentResponse(Box<DuitNowQrCodeResponse>),
    Error(FiuuErrorResponse),
    RecurringResponse(Vec<Box<FiuuRecurringResponse>>),
}
```

**Review by awasthi21 — APPROVED** [review]

File: crates/hyperswitch_connectors/src/connectors/nexixpay.rs [hunk 2]
```diff
@@ -415,10 +426,14 @@ impl ConnectorIntegration<Authorize, PaymentsAuthorizeData, PaymentsResponseData
 
     fn get_url(
         &self,
-        _req: &PaymentsAuthorizeRouterData,
+        req: &PaymentsAuthorizeRouterData,
         connectors: &Connectors,
     ) -> CustomResult<String, errors::ConnectorError> {
-        Ok(format!("{}/orders/3steps/init", self.base_url(connectors)))
+        if req.request.connector_mandate_id().is_none() {
```
Comment (srujanchikke) [inline crates/hyperswitch_connectors/src/connectors/nexixpay.rs:432]: can we check for off_session instead

File: crates/hyperswitch_connectors/src/connectors/nexixpay/transformers.rs [hunk 3]
```diff
@@ -371,53 +435,119 @@ impl TryFrom<&NexixpayRouterData<&PaymentsAuthorizeRouterData>> for NexixpayPaym
...
-                    currency: item.router_data.request.currency,
-                    description: item.router_data.description.clone(),
-                    customer_info,
+        let billing_address = Address {
+            name: item.router_data.get_billing_full_name()?,
+            street: item.router_data.get_billing_line1()?,
+            city: item.router_data.get_billing_city()?,
+            post_code: item.router_data.get_billing_zip()?,
+            country: item.router_data.get_billing_country()?,
+        };
```
Comment (srujanchikke) [inline crates/hyperswitch_connectors/src/connectors/nexixpay/transformers.rs:444]: All required fields should be added in payment_connector_required_fields.rs file

File: crates/hyperswitch_connectors/src/connectors/nexixpay/transformers.rs [hunk 4]
```diff
@@ -371,53 +435,119 @@ impl TryFrom<&NexixpayRouterData<&PaymentsAuthorizeRouterData>> for NexixpayPaym
...
+        let customer_info = CustomerInfo {
+            card_holder_name: item.router_data.get_billing_full_name()?,
+            billing_address: billing_address.clone(),
+            shipping_address: Some(billing_address),
```
Comment (srujanchikke) [inline crates/hyperswitch_connectors/src/connectors/nexixpay/transformers.rs:448]: why are we hardcoding shipping address to billing_address ?

**Review by srujanchikke — COMMENTED** [review]
Please add cypress test cases as well.

**Comment by mrudulvajpayee4935** [comment]
> Please add cypress test cases as well.
Could not run cypress tests since we don't have support for 3ds mandates cypress tests. cc: @likhinbopanna

**Review by srujanchikke — APPROVED** [review]

**Review by awasthi21 — APPROVED** [review]

**Review by awasthi21 — APPROVED** [review]

**Review by srujanchikke — DISMISSED** [review]

**Comment by Gnanasundari24** [comment]
@mrudulvajpayee4935 Please run the cypress testcases for nexixpay

**Comment by mrudulvajpayee4935** [comment]
> have support for 3ds mandates cypress tests.

do we have support for 3ds mandates cypress tests? @Gnanasundari24

File: cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js [hunk 5]
```diff
@@ -16,19 +16,99 @@ const successfulThreeDSTestCardDetails = {
...
 export const connectorDetails = {
   card_pm: {
     PaymentIntent: {
       Request: {
         currency: "EUR",
-        amount: 3545,
+        amount: 6500,
         customer_acceptance: null,
         setup_future_usage: "on_session",
+        billing: {
```
Comment (likhinbopanna) [inline cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js:78]: Can you declare billing as const and use it everywhere else the billing address are same

**Review by awasthi21 — APPROVED** [review]

File: cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js [hunk 6]
```diff
@@ -184,5 +296,332 @@ export const connectorDetails = {
...
+        payment_method_data: {
+          card: successfulThreeDSTestCardDetails,
+        },
+        currency: "EUR",
+        mandate_data: multiUseMandateData,
+        billing: billingAddress,
+      },
+      Response: {
+        status: 200,
+        trigger_skip: true,
```
Comment (pixincreate) [inline cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js:311]: You will have to change the implementation once https://github.com/juspay/hyperswitch/pull/6588 gets merged here and else where for `trigger_skip`

File: cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js [hunk 7]
```diff
@@ -1,3 +1,13 @@
+import { getCustomExchange } from "./Commons";
```
Comment (pixincreate) [inline cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js:1]: can you check the lints?

File: cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js [hunk 8]
```diff
@@ -6,43 +16,119 @@ const successfulThreeDSTestCardDetails = {
...
+  card: {
+    last4: "1111",
+    card_type: null,
+    card_network: null,
+    card_issuer: null,
+    card_issuing_country: null,
+    card_isin: "411111",
+    card_extended_bin: null,
+    card_exp_month: "08",
+    card_exp_year: "25",
```
Comment (pixincreate) [inline cypress-tests/cypress/e2e/PaymentUtils/Nexixpay.js:58]: increase this year to a later date. august 2025 will come sooner than you might expect.

**Review by pixincreate — COMMENTED** [review]
other than that, looks good to me.

i also recommend you to run tests locally to see if everything works.

**Review by likhinbopanna — APPROVED** [review]

**Review by awasthi21 — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**